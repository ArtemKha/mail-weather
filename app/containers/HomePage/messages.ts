/*
 * HomePage Messages
 *
 * This contains all the text for the HomePage container.
 */
import { defineMessages } from 'react-intl';

export const scope = 'containers.HomePage';

export default defineMessages({
  placeholder: {
    id: `${scope}.placeholder`,
    defaultMessage: 'Search',
  },
  pm: {
    id: `${scope}.pm`,
    defaultMessage: 'PM 10',
  },
  wind: {
    id: `${scope}.wind`,
    defaultMessage: 'Ветер',
  },
  temperature: {
    id: `${scope}.temperature`,
    defaultMessage: 'Температура',
  },
  warmer: {
    id: `${scope}.warmer`,
    defaultMessage: 'Теплее чем',
  },
});
