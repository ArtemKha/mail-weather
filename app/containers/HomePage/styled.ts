import styled from 'styles/styled-components';

export const Background = styled.div`
  /*  */
`;

export const SliderSection = styled.section`
  /*  */
`;

export const Container = styled.main`
  max-width: 600px;
  margin: 10vh auto 5vh;
  padding: 10px;
  border-radius: 10px;
  display: flex;
  flex-direction: column;
`;
