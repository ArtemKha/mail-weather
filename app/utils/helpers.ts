import cyrillicToTranslit from 'cyrillic-to-translit-js';
import { City, ResponseCity } from 'containers/App/types';

export function mapResponse(list: ResponseCity[]): City[] {
  return list.map(city => ({
    name: new cyrillicToTranslit().reverse(city.name),
    pm: Math.floor(Math.random() * 200),
    wind: city.wind.speed,
    temp: Math.round(city.main.temp),
  }));
}
