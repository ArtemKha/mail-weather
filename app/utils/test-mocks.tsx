import React from 'react';

export const selectMockComponent = ({ options, value = '', onSelect }) => {
  function handleChange(event) {
    onSelect(event.currentTarget.value);
  }
  return (
    <select data-testid="search" value={value} onChange={handleChange}>
      {options.map(({ name }) => (
        <option key={name} value={name}>
          {name}
        </option>
      ))}
    </select>
  );
};
