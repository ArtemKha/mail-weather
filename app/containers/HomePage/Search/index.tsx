import React, { useState } from 'react';
import { Select } from 'antd';
import { City } from 'containers/App/types';
import { FormattedMessage } from 'react-intl';
import messages from '../messages';

const { Option } = Select;

interface Props {
  options: City[];
  onSelect: (name: string) => void;
}
const Search: React.FC<Props> = ({ options, onSelect }) => {
  const [selected, setSelected] = useState('');
  function filterOption(input, option) {
    return (
      option.props
        .children!.toString()
        .toLowerCase()
        .indexOf(input.toLowerCase()) >= 0
    );
  }

  return (
    <Select
      showSearch
      style={{ width: '100%' }}
      placeholder={<FormattedMessage {...messages.placeholder} />}
      optionFilterProp="children"
      onChange={onSelect}
      value={undefined}
      notFoundContent={<FormattedMessage {...messages.notFound} />}
      filterOption={filterOption}
    >
      {options.map(city => (
        <Option key={city.name} value={city.name}>
          {city.name}
        </Option>
      ))}
    </Select>
  );
};

export default Search;
