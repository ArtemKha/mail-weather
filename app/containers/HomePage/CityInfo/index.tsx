import React from 'react';
import { City } from 'containers/App/types';
import { Container, Info, InfoRow, Title } from './styled';
import { FormattedMessage } from 'react-intl';
import messages from '../messages';

interface Props {
  city: City;
}
export const CityInfo: React.FC<Props> = ({ city }) => {
  return (
    <Container>
      <Title>{city.name}</Title>
      <Info>
        <InfoRow>
          <FormattedMessage {...messages.pm} />: {city.pm}
        </InfoRow>
        <InfoRow>
          <FormattedMessage {...messages.wind} />: {city.wind}
        </InfoRow>
        <InfoRow>
          <FormattedMessage {...messages.temp} />: {city.temp}
        </InfoRow>
      </Info>
    </Container>
  );
};
