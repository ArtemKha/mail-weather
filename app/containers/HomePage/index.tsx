import React, { useState, useEffect } from 'react';
import {
  useSelector as useReduxSelector,
  useDispatch,
  TypedUseSelectorHook,
} from 'react-redux';
import FlipMove from 'react-flip-move';
import { Slider } from 'antd';
import { FormattedMessage } from 'react-intl';
import { ApplicationRootState } from 'types';
import { City } from 'containers/App/types';
import { setTemperature } from 'containers/App/actions';
import {
  Container,
  Section,
  Label,
  RangeInput,
  Filter,
  SearchSection,
} from './styled';
import { CityInfo } from './CityInfo';
import messages from './messages';
import Search from './Search';

const useSelector: TypedUseSelectorHook<
  ApplicationRootState
> = useReduxSelector;

const HomePage: React.FC = () => {
  const [cities, temperature]: [City[], number] = useSelector(({ global }) => [
    global.cities,
    global.temperature,
  ]);

  const [activeCities, setActiveCities] = useState(cities.slice(5, 8));
  const [optionCities, setOptionCities] = useState(cities);

  const dispatch = useDispatch();

  useEffect(() => {
    setOptionCities(cities.filter(city => city.temp >= temperature));
  }, [temperature]);

  function onSelect(name) {
    const city = optionCities.find(city => city.name === name)!;
    if (city) {
      const newCities = new Set(activeCities).add(city);
      setActiveCities(Array.from(newCities));
    }
  }

  function onClose(name) {
    const newCities = activeCities.filter(city => city.name !== name)!;
    setActiveCities(newCities);
  }

  function onFilter(value) {
    dispatch(setTemperature(value));
  }

  return (
    <Container>
      <SearchSection>
        <Search options={optionCities} onSelect={onSelect} />
      </SearchSection>
      <Filter>
        <Label>
          <FormattedMessage {...messages.warmer} />
        </Label>
        <RangeInput>
          <Slider onChange={onFilter} min={-20} defaultValue={0} max={20} />
        </RangeInput>
      </Filter>
      <Section>
        <FlipMove>
          {activeCities
            .filter(city => city.temp >= temperature)
            .map(city => (
              <CityInfo key={city.name} city={city} onClose={onClose} />
            ))}
        </FlipMove>
      </Section>
    </Container>
  );
};

export default HomePage;
