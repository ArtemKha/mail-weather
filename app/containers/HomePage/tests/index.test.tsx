import React from 'react';
import { mocked } from 'ts-jest/utils';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import HomePage from '../index';
import { Provider } from 'react-redux';
import configureStore from 'configureStore';
import history from 'utils/history';
import { ConnectedRouter } from 'connected-react-router';
import { selectMockComponent } from 'utils/test-mocks';
import Search from 'containers/HomePage/Search';
const store = configureStore({}, history);

jest.mock('containers/App/actions');
jest.mock('containers/HomePage/Search');

const getComponent = () =>
  render(
    // tslint:disable-next-line: jsx-wrap-multiline
    <Provider store={store}>
      <IntlProvider locale="en">
        <ConnectedRouter history={history}>
          <HomePage />
        </ConnectedRouter>
      </IntlProvider>
    </Provider>,
  );

describe('<HomePage />', () => {
  afterEach(cleanup);

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = getComponent();
    expect(firstChild).toMatchSnapshot();
  });

  it('should add cities on select input', () => {
    // arrange
    const value = 'Серпухов';

    mocked(Search).mockImplementation(selectMockComponent);

    // act
    const { getByTestId, getAllByTestId } = getComponent();
    const select = getByTestId('search');
    fireEvent.change(select, {
      target: { value },
    });
    const activeCities = getAllByTestId('list-city-name');

    // assert
    expect(
      activeCities.some(element => element.textContent!.includes(value)),
    ).toBe(true);
  });

  it('should filter active cities on a temperature filter move', () => {
    // coming soon (not really)
  });

  it('should remove an active city on a remove button push', () => {
    // coming soon (not really)
  });
});
