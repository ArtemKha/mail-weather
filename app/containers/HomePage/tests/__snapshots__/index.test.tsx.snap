// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<main
  class="sc-EHOje gpidVX"
>
  <section
    class="sc-bwzfXH yQUbn"
  >
    <div
      class="ant-select ant-select-enabled"
      style="width: 100%;"
    >
      <div
        aria-autocomplete="list"
        aria-controls="test-uuid"
        aria-expanded="false"
        aria-haspopup="true"
        class="ant-select-selection
            ant-select-selection--single"
        role="combobox"
        tabindex="0"
      >
        <div
          class="ant-select-selection__rendered"
        >
          <div
            class="ant-select-selection__placeholder"
            style="display: block; user-select: none;"
            unselectable="on"
          >
            <span>
              Выберете город
            </span>
          </div>
          <div
            class="ant-select-search ant-select-search--inline"
            style="display: none;"
          >
            <div
              class="ant-select-search__field__wrap"
            >
              <input
                autocomplete="off"
                class="ant-select-search__field"
                value=""
              />
              <span
                class="ant-select-search__field__mirror"
              >
                
                 
              </span>
            </div>
          </div>
        </div>
        <span
          class="ant-select-arrow"
          style="user-select: none;"
          unselectable="on"
        >
          <i
            aria-label="icon: down"
            class="anticon anticon-down ant-select-arrow-icon"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </i>
        </span>
      </div>
    </div>
  </section>
  <section
    class="sc-htpNat boLlBa"
  >
    <label
      class="sc-bxivhb cRHmGX"
    >
      <span>
        Теплее чем
      </span>
    </label>
    <label
      class="sc-ifAKCX IgzzV"
    >
      <div
        class="ant-slider"
      >
        <div
          class="ant-slider-rail"
        />
        <div
          class="ant-slider-track"
          style="left: 0%; width: 50%;"
        />
        <div
          class="ant-slider-step"
        />
        <div
          aria-disabled="false"
          aria-valuemax="20"
          aria-valuemin="-20"
          aria-valuenow="0"
          class="ant-slider-handle"
          role="slider"
          style="left: 50%; transform: translateX(-50%);"
          tabindex="0"
        />
        <div
          class="ant-slider-mark"
        />
      </div>
    </label>
  </section>
  <section
    class="sc-bdVaJa fGyBAk"
  >
    <div
      style="position: relative;"
    />
  </section>
</main>
`;
